lib/charms/tls_certificates_interface/v4/tls_certificates.py

```diff
@@ -1114,45 +1147,79 @@ def _get_app_or_unit(self) -> Union[Application, Unit]:
     @property
     def private_key(self) -> Optional[PrivateKey]:
         """Return the private key."""
+        if self._private_key:
+            return self._private_key
         if not self._private_key_generated():
             return None
         secret = self.charm.model.get_secret(label=self._get_private_key_secret_label())
         private_key = secret.get_content(refresh=True)["private-key"]
         return PrivateKey.from_string(private_key)
 
-    def _generate_private_key(self) -> None:
+    def _ensure_private_key(self) -> None:
+        """Make sure there is a private key to be used.
+
+        It will make sure there is a private key passed by the charm using the private_key
+        parameter or generate a new one otherwise.
+        """
+        if self._private_key:
+            self._remove_private_key_secret()
+            return
         if self._private_key_generated():
+            logger.debug("Private key already generated")
             return
-        private_key = generate_private_key()
-        self.charm.unit.add_secret(
-            content={"private-key": str(private_key)},
-            label=self._get_private_key_secret_label(),
-        )
-        logger.info("Private key generated")
+        self._generate_private_key()
 
     def regenerate_private_key(self) -> None:
         """Regenerate the private key.
 
         Generate a new private key, remove old certificate requests and send new ones.
+
+        Raises:
+            TLSCertificatesError: If the private key is passed by the charm using the
+                private_key parameter.
         """
+        if self._private_key:
+            raise TLSCertificatesError(
+                "Private key is provided by the charm through the private_key parameter, "
+                "this function can't be used"
+            )
         if not self._private_key_generated():
             logger.warning("No private key to regenerate")
             return
-        self._regenerate_private_key()
+        self._generate_private_key()
         self._cleanup_certificate_requests()
         self._send_certificate_requests()
 
-    def _regenerate_private_key(self) -> None:
-        secret = self.charm.model.get_secret(label=self._get_private_key_secret_label())
-        secret.set_content({"private-key": str(generate_private_key())})
-        secret.get_content(refresh=True)
+    def _generate_private_key(self) -> None:
+        """Generate a new private key."""
+        self._store_private_key_in_secret(generate_private_key())
+        logger.info("Private key generated")
 
     def _private_key_generated(self) -> bool:
```

Comment:
A docstring might be helpful on this function, because the name is a little ambiguous. I was getting a little confused with `_private_key`, `_private_key_generated`, `_generate_private_key`. Specfically, emphasizing we have `_private_key` for provided key, and `_private_key_generated` for one that was generated by this library is worth highlighting somewhere.

Reply:
Docstring improved. 
I tried to think of some clearer naming. But it always goes back to `regenerate_private_key` which is a public function, so we can't change its name and at the same time I don't want to introduce new terms for the internals of the lib.
Happy to listen to any ideas.